adafruit_logger.py

```diff
@@ -0,0 +1,171 @@
+# THE SOFTWARE.
+"""
+`adafruit_logger`
+================================================================================
+
+Logging module for CircuitPython
+
+
+* Author(s): Dave Astels
+
+Implementation Notes
+--------------------
+
+**Hardware:**
+
+
+**Software and Dependencies:**
+
+* Adafruit CircuitPython firmware for the supported boards:
+  https://github.com/adafruit/circuitpython/releases
+
+"""
+#pylint:disable=redefined-outer-name,consider-using-enumerate,no-self-use
+
+# imports
+
+__version__ = "0.0.0-auto.0"
+__repo__ = "https://github.com/adafruit/Adafruit_CircuitPython_Logger.git"
+
+import time
+
+LEVELS = [(00, 'NOTSET'),
+          (10, 'DEBUG'),
+          (20, 'INFO'),
+          (30, 'WARNING'),
+          (40, 'ERROR'),
+          (50, 'CRITICAL')]
+
+for value, name in LEVELS:
+    globals()[name] = value
+
+def level_for(value):
+    """Convert a numberic level to the most appropriate name.
+    value -- a numeric level
+    """
+    for i in range(len(LEVELS)):
+        if value < LEVELS[i][0]:
+            return LEVELS[i-1][1]
+    return LEVELS[0][1]
+
+class LoggingHandler(object):
+    """Abstract logging message handler."""
+
+    def format(self, level, msg):
+        """Generate a timestamped message.
+        level -- the logging level
+        msg -- the message to log
+        """
+        now = time.localtime()
+        time_vals = (now.tm_year, now.tm_mon, now.tm_mday,
+                     now.tm_hour, now.tm_min, now.tm_sec)
+        timestamp = '%4d/%02d/%02d %02d:%02d:%02d' % time_vals
+        return '{0}: {1} - {2}'.format(timestamp, level_for(level), msg)
+
+    def emit(self, level, msg):
+        """Send a message where it should go.
+        Place holder for subclass implementations.
+        """
+        raise NotImplementedError()
+
+
+class PrintHandler(LoggingHandler):
+    """Send logging messages to the console by using print."""
+
+    def emit(self, level, msg):
+        """Send a message to teh console.
+        level -- the logging level
+        msg -- the message to log
+        """
+        print(self.format(level, msg))
+
+
+# The level module-global variables get created when loaded
+#pylint:disable=undefined-variable
+
+class Logger(object):
+    """Provide a logging api."""
+
+    def __init__(self, handler=None):
+        """Create an instance.
+        handler -- what to use to output messages. Defaults to a PrintHandler.
+        """
+        self._level = NOTSET
+        if handler is None:
+            self._handler = PrintHandler()
+        else:
+            self._handler = handler
+
+    @property
+    def level(self):
+        """Get the level."""
+        return self._level
+
+    @level.setter
+    def level(self, value):
+        """Set the level."""
+        self._level = value
+
+    def log(self, level, format_string, *args):
+        """Log a message.
+        level -- the priority level at which to log
```

Comment:
Please use `:param level: The priority level at which to log` formatting for documenting keywords. Include whitespace above the first `:param` and below the last `:param`, but it is not necessary between `:params`.
e.g.
```
"""Log a message.

:param level: The priority of the level at which to log.
:param format_string: the core message string with embedded formatting directives
:param args: arguments ``format_string.format()``, can be empty

"""
```